collaboration/sig-core/proposals/kyma-controller-manager.md

```diff
@@ -0,0 +1,181 @@
+
+Created on 2019-10-11 by Antoine Cotten (@antoineco).
+
+## Status
+
+DRAFT. Proposed on 2019-10-11.
+
+## Motivation
+
+Kyma is composed of multiple components that share a lot of architectural patterns, yet are written in isolation
+following different engineering standards and using heterogeneous SDKs and libraries, before eventually being released
+as _one product_.
+
+Most of these components involve one or more Kubernetes controllers that synchronize API objects using identical
+building blocks: watches, work queues, reconciliation loops, etc. Every controller instance maintains its own local
+cache of several Kubernetes resources, which can result in a significant memory footprint in productive environments.
+Besides, each of these controllers is operated in a similar fashion by nature: as a single instance or in an
+active-passive setup.
+
+These similarities represent an opportunity for us to consolidate Kyma controllers and run them under the umbrella of a
+single _manager_, so that they can reuse common libraries, maintain a consistent level of compatibility with the
+underlying Kubernetes layer, share memory efficiently, and avoid the fragmentation we are currently observing in the
+project.
+
+This mode of operation is directly inspired by the Kubernetes [controller-manager][k8s-cm].
+
+## Goals
+
+This initiative aims at improving three aspects:
+
+**Maintainability**
+
+Streamline the codebase regarding controllers to harmonize the different implementation aspects. (_"Common opinion"_)
+
+**Footprint**
+
+Reduce the resource consumption by sharing internal API object caches whenever possible. (_"Store once"_)
+
+**Operability**
+
+Improve operational aspects by reducing the number of controllers that have to be deployed separately. (_"Run once"_)
+
+## Constraints
+
+It must be possible to opt out of each individual controller inside the manager to keep Kyma modular. This could be
+achieved using startup flags and/or a ConfigMap.
+
+A controller ("operator") SDK should be agreed upon by all code owners. [kyma#4327][kyma4327] initiated
+a discussion around that topic, but no conclusion has been reached at the time of writing. For the initial version of
+this document, we assume that [`kubebuilder`][kb] will be our SDK of choice. A few reasons to believe this assumption is
+safe to make:
+- well maintained and documented by a dedicated group (SIG) within the Kubernetes project
+- leverages `controller-runtime`, a collection of sensible libraries with everything a controller typically needs
+- bootstaps projects with the manager pattern enabled right from the start
+- includes tooling oriented towards future extensibility, with helpers to generate new APIs
+- RBAC objects are part of the code generation
+
+## Proposed solution
+
+### Project structure
+
+The structure of the controller manager is initialized from _scratch_, existing Kyma components get incorporated
+gradually into it after suitable candidates have been identified.
```

Comment:
Who would own that project the codebase for that controller manager? Right now the different controllers are owned by individual teams, who are managing the lifecycle of that component. How are we planning to coordinate that if we have a central controller manager?

Reply:
Totally agree. The teams will still own their APIs and controllers, but we need some governance to ensure all code owners are on the same page when it comes to project structure, dependencies management, etc.